openedx/core/djangoapps/content/course_overviews/management/commands/backfill_history.py

```diff
@@ -0,0 +1,138 @@
+            type=int,
+            help="Maximum number of bytes to read from input file in each batch.",
+        )
+
+        parser.add_argument(
+            "--input_root",
+            action="store",
+        )
+
+    def handle(self, *args, **options):
+        byte_size = options['size']
+        sleep_between = options['sleep_between']
+        input_root = options['input_root']
+
+        for table_info in self.TABLES:
+            table = table_info['name']
+            historical_table = "_historical".join(table.rsplit('_', 1))
+            exclude_column = table_info['exclude_column']
+            input_filename = table_info['input_filename']
+            file_path = os.path.join(input_root, input_filename)
+
+            with connection.cursor() as cursor:
+                query = u"""
+                    SELECT
+                        column_name
+                    FROM information_schema.columns
+                    WHERE table_name='{}'
+                    ORDER BY ordinal_position
+                    """.format(table)
+                cursor.execute(query)
+                columns = [column[0] for column in cursor.fetchall()]
+            if exclude_column in columns:
+                columns.remove(exclude_column)
+            columns = ['`{}`'.format(c) for c in columns]
+
+            with open(file_path) as input_file:
+                # Skip the header
+                input_file.readline()
+                while True:
+                    lines = input_file.readlines(byte_size)
+                    if not lines:
+                        break
+
+                    values = [line.strip('\n').split(',') for line in lines]
+                    # Add history columns data
+                    [value.extend([self.DATE, self.HISTORY_CHANGE_REASON, '+', self.HISTORY_USER_ID]) for value in values]
+                    # Convert to tuple
+                    values = [tuple(value) for value in values]
+                    # Get the ids
+                    ids = [value[0] for value in values]
+
+                    # Checks for existing historical records
+                    with connection.cursor() as cursor:
+                        query = u"""
+                            SELECT COUNT(1)
+                            FROM {historical_table}
+                            WHERE ID in ({ids})
+                            AND history_type='+'
+                            """.format(
+                                historical_table=historical_table,
+                                ids=','.join(ids)
+                            )
+                        log.info(query)
+                        cursor.execute(query)
+                        count = cursor.fetchone()[0]
+
+                    if count==len(ids):
+                        log.info(u"Initial history records already exist for ids %s..%s - skipping.", ','.join(ids[:2]), ','.join(ids[-2:]))
+                        continue
+                    elif count!=0:
```

Comment:
If I'm reading this section correctly, if _all_ of the rows in the current batch have pre-existing history we will log and continue, but if only some of them do we will fail? The latter seems like a much more likely case, and we would have to re-generate the input file to work around it to continue, wouldn't we? It seems like we would just want to skip those rows if they exist (ex. select the ids of those rows instead of just the count, then just skip over them in the loop below).

Reply:
The input file that would be passed in would contain all the rows that need a historical record. 
As we're doing the insert atomically, the rows will either exist or wouldn't. 

In case there is a partial match(which really shouldn't happen), it warrants an investigation.

Reply:
I guess the longer between when the input file is generated and when the script is run increases the likelihood of partial matches? Especially on something like `entitlements_courseentitlement` which I think can have changes written pretty erratically (and so new history created) even in old rows. I don't think it hurts anything to leave it as it is, but I wouldn't be surprised if we have a partial match if the file is generated more than a day or so before the script is run.

Reply:
@bmedx I don't expect this to happen. This management command will only create historical records for rows which don't have a corresponding historical record with '+' as history_type.